trace-obfuscation/src/replacer.rs

```diff
@@ -0,0 +1,188 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present
+// Datadog, Inc.
+
+use datadog_trace_protobuf::pb;
+use regex::Regex;
+
+#[derive(Debug)]
+pub struct ReplaceRule<'a> {
+    // name specifies the name of the tag that the replace rule addresses. However,
+    // some exceptions apply such as:
+    // * "resource.name" will target the resource
+    // * "*" will target all tags and the resource
+    name: &'a str,
+
+    // re holds the regex pattern for matching.
+    re: regex::Regex,
+
+    // repl specifies the replacement string to be used when Pattern matches.
+    repl: &'a str,
+}
+
+/// replace_trace_tags replaces the tag values of all spans within a trace with a given set of rules.
+pub fn replace_trace_tags(trace: &mut [pb::Span], rules: &[ReplaceRule]) {
```

Comment:
Might be worth addin a criterion benchmark for this function.

In near feature I plan to add the benchmarking platform to the CI config. So that we'll get PR comments if a regression gets detected

Reply:
added a new criterion bench for `replace_trace_tags` which can be run with `cargo bench`